### PR TITLE
Mark remote cache errors as retriable by expanding `--incompatible_remote_use_new_exit_code_for_lost_inputs`

### PR DESCRIPTION
**Pull request by criemen**
If `--incompatible_remote_use_new_exit_code_for_lost_inputs` is set, or if `--experimental_remote_cache_eviction_retries` is set, all remote cache errors during cache lookup will be (eligible for) being retried by restarting the build. This does not apply to remote cache errors during remote execution, though.

Fixes https://github.com/bazelbuild/bazel/issues/23033.

Some words about the design and the backwards compatibility of this change.
I'm not super happy with where I ended up here:
I'm using `REMOTE_CACHE_EVICTED` here for errors that are not cache evictions. The question is, what is more disruptive to bazel users - return exit code 39 more often for remote cache errors (not just evictions), which is the approach taken in this PR, or introduce yet another `incompatible` flag that introduces, say, exit code 40 for remote cache errors that are not eviction errors (or all remote cache errors, if we then merge eviction+general errors).
We can easily change the semantics of `experimental_remote_cache_eviction_retries` as it's experimental, but with `incompatible_remote_use_new_exit_code_for_lost_inputs` this is maybe more sketchy (but done in this PR!).

On the other hand, I don't know the stability guarantees about the protobuf API - `Spawn.Code.REMOTE_CACHE_FAILED` would be the perfect to use for this case, but it's associated with exit code 34, which is all sorts of remote (execution) errors that we, as far as I can see, don't necessarily want to retry when the user specifies `experimental_remote_cache_eviction_retries` (otherwise the flag would change meaning a lot).
Could/should we introduce `REMOTE_CACHE_FAILED_NEW` with exit code 40 that then gets used if the user specifies an incompatible flag? That means that users would need to adjust all their infra scripts to move from exit code 39 (or 34) to 40 at some point, which is a lot of churn.
Could we instead rename `REMOTE_CACHE_EVICTED` to `REMOTE_CACHE_FAILED_NEW`, or does this violate some protobuf longevity guarantees?

Writing all of this down, I'd ideally like to rename `REMOTE_CACHE_EVICTED` to `REMOTE_CACHE_FAILED_NEW`, and then, depending on bazel policy, have `incompatible_remote_use_new_exit_code_for_lost_inputs` emit these errors more often, or introduce another incompatible flag to move more remote errors from `REMOTE_CACHE_FAILED` to `REMOTE_CACHE_FAILED_NEW`.

There's also another problem this PR doesn't address: Remote cache lookups during remote execution - in `RemoteSpawnRunner`, we could track more closely if an exception stems from remote cache lookup, and then follow the semantics outlined above for that - right now, we only distinguish between cache-not-found errors and everything else. The reason I didn't work on this is that if your remote cache is unstable during remote execution, you have a bigger problem anyway.
While you might hit #23033 during a local fallback execution, I believe this PR already takes care of that code path.
For all other remote cache errors during remote execution, I don't see how retrying the entire build will help.

This is a lot of questions I don't feel comfortable answering myself, as I don't know exactly what stability guarantees bazel provides to its users. I'm happy to rework this PR based on feedback, though.

If this PR is accepted eventually, it'd be great to get this in for 7.3.0, if there's enough time, and the PR isn't too breaking.

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/exec/ExecutionOptions.java [hunk 1]
```diff
@@ -534,24 +534,29 @@ public boolean usingLocalTestJobs() {
...
+      // TODO: when this flag is moved to non-experimental, rename it to a more general name
+      // to reflect the new logic - it's not only about cache evictions.
       name = "experimental_remote_cache_eviction_retries",
       defaultValue = "0",
       documentationCategory = OptionDocumentationCategory.REMOTE,
       effectTags = {OptionEffectTag.EXECUTION},
       help =
-          "The maximum number of attempts to retry if the build encountered remote cache eviction"
-              + " error. A non-zero value will implicitly set"
+          "The maximum number of attempts to retry if the build encountered a remote cache"
```
Comment (coeuvre) [inline src/main/java/com/google/devtools/build/lib/exec/ExecutionOptions.java:549]: I would rephrase to:

> The maximum number of attempts to retry if the build encountered a transient remote cache error that would otherwise fail the build. Applies for example when artifacts are evicted from the remote cache, or in certain cache failure conditions. A non-zero value ...

- Emphasize `transient` remote cache error.
- Do not mention `--remote_download_minimal` because they are orthogonal.

File: src/main/java/com/google/devtools/build/lib/runtime/BlazeCommandDispatcher.java [hunk 2]
```diff
@@ -365,7 +365,7 @@ private BlazeCommandResult execExclusively(
                 env.getCommandId()));
         return Preconditions.checkNotNull(lastResult);
       } else {
-        outErr.printErrLn("Found remote cache eviction error, retrying the build...");
+        outErr.printErrLn("Found fatal remote cache error, retrying the build...");
```
Comment (coeuvre) [inline src/main/java/com/google/devtools/build/lib/runtime/BlazeCommandDispatcher.java:368]: Rename `fatal` to `transient`.

File: src/main/protobuf/failure_details.proto [hunk 3]
```diff
@@ -227,6 +227,8 @@ message Spawn {
     //   refactored to prohibit undetailed failures
     UNSPECIFIED_EXECUTION_FAILURE = 12 [(metadata) = { exit_code: 1 }];
     FORBIDDEN_INPUT = 13 [(metadata) = { exit_code: 1 }];
+    // TODO: This also includes other remote cache errors, not just evictions,
```
Comment (coeuvre) [inline src/main/protobuf/failure_details.proto:230]: Rephrase the comment to:

> Note: This also includes other remote cache errors, not just evictions.
> TODO: Rename it to a more general name when `--experimental_remote_cache_eviction_retries` is moved to non-experimental.

File: src/main/java/com/google/devtools/build/lib/exec/AbstractSpawnStrategy.java [hunk 4]
```diff
@@ -306,7 +306,18 @@ public ListenableFuture<Void> prefetchInputs() throws ForbiddenActionInputExcept
                         .setSpawn(FailureDetails.Spawn.newBuilder().setCode(code))
                         .build());
               } else {
-                throw e;
+                if (executionOptions.remoteRetryOnCacheError > 0) {
```
Comment (coeuvre) [inline src/main/java/com/google/devtools/build/lib/exec/AbstractSpawnStrategy.java:309]: Because this is an incompatible change, I would also like to guard it with `executionOptions.useNewExitCodeForLostInputs`. Then we have the same if condition, maybe change the code to:

```
if (executionOptions.useNewExitCodeForLostInputs || executionOptions.remoteRetryOnCacheError > 0) {
    if (BulkTransferException.allCausedByCacheNotFoundException(e)) {
        throw new EnvironmentalExecException(
                    e,
                    FailureDetail.newBuilder()
                        .setMessage("Failed to fetch blobs because they do not exist remotely.")
                        .setSpawn(FailureDetails.Spawn.newBuilder().setCode(Code.REMOTE_CACHE_EVICTED))
                        .build());
    } else {
        throw new EnvironmentalExecException(
                      e,
                      FailureDetail.newBuilder()
                          .setMessage("Failed to fetch blobs because from the remote cache.")
                          .setSpawn(
                              FailureDetails.Spawn.newBuilder().setCode(Code.REMOTE_CACHE_EVICTED))
                          .build());
    }
} else {
    throw e;
}
```

**Review by coeuvre — CHANGES_REQUESTED** [review]
Thanks for the PR and sorry for the delay (I was on vacation)!

I like your design and where the PR is moving to: when fixing issues, we want to keep backward compatibility as much as possible without introducing new incompatible flag.

Generally speaking, a build could encounter two types of remote cache errors: the permanent error that retries cannot help, such as wrong configuration, and the transient error that retries could fix, e.g. remote cache eviction, remote cache is temporarily down and etc. I think the `REMOTE_CACHE_FAILED (34)` could be used for the former one and `REMOTE_CACHE_EVICTED (39)` could be used for the latter one.

I left some comments but otherwise LGTM.

File: src/main/java/com/google/devtools/build/lib/exec/AbstractSpawnStrategy.java (hunk 4 above)
Comment (criemen) [reply src/main/java/com/google/devtools/build/lib/exec/AbstractSpawnStrategy.java:309]: That's not equivalent, and lacking the case where the legacy code path throws the `EnvironmentalExecException`.
I'll push a new version of the if though that has the experimental case and the legacy case clearly separated, so that deleting the legacy branch is very easy once cleaning up this experimental flag.

**Comment by criemen** [comment]
I've force-pushed a rebase onto master, but the commits are the same as before, and I've addressed the review comments in a separate commit to make it easier for you to verify the changes.

Sorry about the delay, I went on PTO just as your review came in, and catching up took me a while.

**Review by coeuvre — APPROVED** [review]
Thanks for the PR! Importing.

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.4.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.4.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.4.0rc1`. Thanks!
